### PR TITLE
Fix missing PR comments

### LINKED ISSUE #12515: Review commets on PR 11877
#### Problem
Apparently my review on #11877 never got submitted, so those comments are not addressed in the code that landed.  One of them is a critical security bug, so needs to be addressed before we ship.

See https://github.com/project-chip/connectedhomeip/pull/11877#pullrequestreview-808033182

 #### Proposed Solution
Fix bugs.

I apparently can't assign this to @mkardous-silabs.....

### PR DESCRIPTION
**Pull request by mkardous-silabs**
#### Problem
* Fixes #12515 

#### Change overview
* Added endpoint index check in OnOff and ColorControl cluster
* Fixed faulty callback signature

#### Testing
* Automated tests to validate no regressions where caused

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
fast track: pr up for  sufficient time for  cross timezone review.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 1]
```diff
@@ -1813,7 +1917,14 @@ void ColorControlServer::updateXYCommand(EndpointId endpoint)
 ColorControlServer::Color16uTransitionState * ColorControlServer::getTempTransitionState(EndpointId endpoint)
 {
     uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
-    return &colorTempTransitionStates[index];
+
+    Color16uTransitionState * state = nullptr;
+    if (index < EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT)
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1922]:
```suggestion
    if (index < ArraySize(colorTempTransitionStates))
```

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 2]
```diff
@@ -498,8 +505,15 @@ bool OnOffServer::areStartUpOnOffServerAttributesTokenized(EndpointId endpoint)
  */
 EmberEventControl * OnOffServer::getEventControl(EndpointId endpoint)
 {
-    uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, OnOff::Id);
-    return &eventControls[index];
+    uint16_t index            = emberAfFindClusterServerEndpointIndex(endpoint, OnOff::Id);
+    EmberEventControl * event = nullptr;
+
+    if (index < EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT)
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:511]:
```suggestion
    if (index < ArraySize(eventControls))
```

File: src/app/clusters/on-off-server/on-off-server.h [hunk 3]
```diff
@@ -87,6 +87,12 @@ struct OnOffEffect
     OnOffEffect(
         chip::EndpointId endpoint, OffWithEffectTriggerCommand offWithEffectTrigger,
         uint8_t effectIdentifier = static_cast<uint8_t>(EMBER_ZCL_ON_OFF_EFFECT_IDENTIFIER_DELAYED_ALL_OFF),
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.h:89]: What about type-safety for this one, where it _is_ possible?

File: src/app/clusters/on-off-server/on-off-server.h [hunk 4]
```diff
@@ -87,6 +87,12 @@ struct OnOffEffect
...
+
+        /*
+         * effectVariant's type depends on the effect effectIdentifier so we don't know the type at compile time.
+         * Casting to uint8_t for more flexibilty since the type can be OnOffDelayedAllOffEffectVariant or
+         * OnOffDelayedAllOffEffectVariant
+         */
         uint8_t effectVariant = static_cast<uint8_t>(EMBER_ZCL_ON_OFF_DELAYED_ALL_OFF_EFFECT_VARIANT_FADE_TO_OFF_IN_0P8_SECONDS));
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.h:96]: Could probably handle this via chip::Variant...  Followup for that?

Comment (mkardous-silabs) [reply src/app/clusters/on-off-server/on-off-server.h:96]: fixed in #12813

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 2 above)
Comment (mkardous-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:511]: fixed in #12813

File: src/app/clusters/on-off-server/on-off-server.h (hunk 3 above)
Comment (mkardous-silabs) [reply src/app/clusters/on-off-server/on-off-server.h:89]: fixed in #12813

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 1 above)
Comment (mkardous-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:1922]: fixed in #12813

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 5]
```diff
@@ -1518,7 +1587,14 @@ void ColorControlServer::updateHueSatCommand(EndpointId endpoint)
 ColorControlServer::Color16uTransitionState * ColorControlServer::getXTransitionState(EndpointId endpoint)
 {
     uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
-    return &colorXtransitionStates[index];
+
+    Color16uTransitionState * state = nullptr;
+    if (index < EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT)
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:1592]: fixed in #12813

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 6]
```diff
@@ -1530,7 +1606,14 @@ ColorControlServer::Color16uTransitionState * ColorControlServer::getXTransition
 ColorControlServer::Color16uTransitionState * ColorControlServer::getYTransitionState(EndpointId endpoint)
 {
     uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
-    return &colorYtransitionStates[index];
+
+    Color16uTransitionState * state = nullptr;
+    if (index < EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT)
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:1611]: fixed in #12813

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 7]
```diff
@@ -366,8 +383,14 @@ ColorControlServer::ColorHueTransitionState * ColorControlServer::getColorHueTra
  */
 ColorControlServer::Color16uTransitionState * ColorControlServer::getSaturationTransitionState(EndpointId endpoint)
 {
-    uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
-    return &colorSatTransitionStates[index];
+    uint16_t index                  = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
+    Color16uTransitionState * state = nullptr;
+
+    if (index < EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT)
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:389]: fixed in #12813

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 8]
```diff
@@ -354,8 +365,14 @@ bool ColorControlServer::computeNewColor16uValue(ColorControlServer::Color16uTra
  */
 ColorControlServer::ColorHueTransitionState * ColorControlServer::getColorHueTransitionState(EndpointId endpoint)
 {
-    uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
-    return &colorHueTransitionStates[index];
+    uint16_t index                  = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
+    ColorHueTransitionState * state = nullptr;
+
+    if (index < EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT)
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:371]: fixed in #12813

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 9]
```diff
@@ -254,8 +259,14 @@ uint16_t ColorControlServer::computeTransitionTimeFromStateAndRate(ColorControlS
  */
 EmberEventControl * ColorControlServer::getEventControl(EndpointId endpoint)
 {
-    uint16_t index = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
-    return &eventControls[index];
+    uint16_t index            = emberAfFindClusterServerEndpointIndex(endpoint, ColorControl::Id);
+    EmberEventControl * event = nullptr;
+
+    if (index < EMBER_AF_COLOR_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT)
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:265]: fixed in #12813

**Pull request merged**